core/src/smartcontracts/isi/permissions/mod.rs

```diff
@@ -72,6 +125,38 @@ pub enum ValidatorVerdict {
     Allow,
 }
 
+impl PredicateSymbol for ValidatorVerdict {
```

Comment:
this is getting into the realm of ternary logic and a rookie mistake. 

We can't apply boolean arithmetic to ternary constructs, but thankfully (thanks to some abstract logicians) you can decouple ternary logic into binary logic plus extensions. 

### True + False + Skip = (True * Skip) + (False * Skip) 

You want to have two extra methods: `false()`, and `skip()`. So your `is_truthful()` ⇒ 
```rust 
fn true(&self) -> Option<bool> {
	match self {
           Self::Accept(_) => Some(true), 
	   Self::Deny(_) => Some(false), 
           _ => None,
       }
}
```

And then you use the monadic methods on `Option` to track the necessary skipping behaviour. You get rid of the associated `EvaluatesTo` type and **ordinary** rust logic applies. This follows the design of `PartialEq`. 

So in places where you'd use `is_truthful` which is more accurately `not_false`, you'd use `true().unwrap_or(false)`, and similarly for the other method. 

Then you get the short-circuiting and ignoring behaviour for basically free.

Reply:
We cannot get rid of `EvaluatesTo` entirely, since `ValidatorVerdict` contains state we want to preserve.
I've rewritten the logic more clearly with `ControlFlow`, see if it looks better this way.

Reply:
Like I said, the problem is the ternary logic. If you insist on doing it this way, please add tests, to demonstrate that it is indeed a compliant implenentation of https://www.wikidoc.org/index.php/Ternary_logic

Reply:
Ah, okay. I didn't quite understand your suggestion.
`ValidatorVerdict` does not and should not conform to *the* ternary logic as per link you provided, since that would e.g. mean that `Skip or Deny = Skip`, which would be a security issue. I'm not sure if `(Validators, ValidatorVerdicts, Operations, Operators)` conforms to definition of _a_ first-order logic or propositional calculus, but I don't think it is really relevant here; what is relevant is that definitions of `and`, `or` and `not` make sense under the circumstances and conform to De Morgan's laws, which I think they do.
That said, I absolutely agree that it's complex enough logic that it warrants being tested.

Reply:
> it's an overkill, but I think it's important, since iroha security fundamentally depends on Validators.

No. It's not overkill, and it's good that it's written as a test. Ideally  these things are written using a formal verification harness, but in this particular case, we need any violation of the logic to be detectable.